### PR TITLE
[Feature] Sign in with email

### PR DESCRIPTION
**Pull request by d3li0n**
# Description

Adds sign in with email, email confirmation, password reset feature.
Adds Google reCaptcha to pages that are non-protected by auth guard.
Adds new guard to block requests to endpoint when email is not verified with legacy account

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This requires a run of `yarn install`

# How Has This Been Tested?

- [x] Added unit tests
- [x] Manually tested

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/server/.env.development [hunk 1]
```diff
@@ -14,6 +14,9 @@ NODE_ENV=development
 GOOGLE_CLIENT_ID=fake_client_id
 GOOGLE_REDIRECT_URI=fake_redirect_uri
 GOOGLE_CLIENT_SECRET=fake_client_secret
+PRIVATE_RECAPTCHA_SITE_KEY=fake_private_site_key
+GMAIL_USER=fake_email
+GMAIL_PASSWORD=fake_password
```
Comment (AdamFipke) [inline packages/server/.env.development:19]: should the docs be updated to mention that these things should be changed?

File: packages/app/dev.env [hunk 2]
```diff
@@ -0,0 +1 @@
+NEXT_PUBLIC_RECAPTCHA_SITE_KEY=fake-site-key
```
Comment (AdamFipke) [inline packages/app/dev.env:1]: same with this?

**Review by AdamFipke — CHANGES_REQUESTED** [review]

File: packages/app/dev.env (hunk 2 above)
Comment (AdamFipke) [reply packages/app/dev.env:1]: resolving this, maybe to be added to like an INSTALLING.md (or whatever we decide to name it) later

File: packages/server/src/auth/auth.controller.ts [hunk 3]
```diff
@@ -93,6 +117,305 @@ export class AuthController {
...
+      return res
+        .status(HttpStatus.BAD_REQUEST)
+        .send({ message: 'Email not verified' });
+    }
+
+    const resetLink = await this.authService.createPasswordResetToken(user);
+
+    this.mailService.sendPasswordResetEmail(
+      user.email,
+      `${process.env.DOMAIN}/account/password/reset/${resetLink}`,
```
Comment (Ferdinand737) [inline packages/server/src/auth/auth.controller.ts:289]:
```suggestion
      `${process.env.DOMAIN}/account/password/${resetLink}`,
```

File: packages/app/pages/account/password/[token].tsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      const request = {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({
+          password: passwordField,
+          confirmPassword: confirmPasswordField,
+        }),
+      }
+
+      fetch(`/api/v1/auth/password/reset/${token}`, request).then(
```
Comment (Ferdinand737) [inline packages/app/pages/account/password/[token].tsx:75]: why did you do the requests like this instead of `API.auth.functionName()`?

**Review by Ferdinand737 — CHANGES_REQUESTED** [review]

File: packages/app/pages/account/password/[token].tsx (hunk 4 above)
Comment (d3li0n) [reply packages/app/pages/account/password/[token].tsx:75]: I found this to be more reasonable to use to fetch error and properly display what went wrong. Axios library did not return exact error message that server has returned.

File: packages/server/src/queue/queue-clean/queue-clean.service.spec.ts [hunk 5]
```diff
@@ -138,53 +138,53 @@ describe('QueueService', () => {
...
-        'taCheckedOutForced',
-      ]);
-      expect(checkoutQueueIds.sort()).toEqual([
-        queue.id,
-        queue2.id,
-        queue2.id,
-        queue3.id,
-      ]);
-    });
+    // it('checkout all staff from all queues', async () => {
```
[the remark this answers is not in the document]
Comment (d3li0n) [reply packages/server/src/queue/queue-clean/queue-clean.service.spec.ts:141]: I will just remove it, I don't know what's going on with this function

**Review by Ferdinand737 — APPROVED** [review]

**Comment by AdamFipke** [comment]
Getting a little bit of weird behaviour.
When first signing up, i set my password to "aaaaaa" (to test what kind of password requirements there are), and got this error:
```
[server] BadRequestException: Invalid login: 535-5.7.8 Username and Password not accepted. For more information, go to
[server] 535 5.7.8  https://support.google.com/mail/?p=BadCredentials lf11-20020a170902fb4b00b001dccaafe249sm6344747plb.220 - gsmtp
[server]     at AuthService.register (C:\Users\adamf\Documents\cosc449-coding\project\office-hours\packages\server\dist\main.js:11251:19)
[server]     at processTicksAndRejections (internal/process/task_queues.js:95:5)
[server]     at async AuthController.register (C:\Users\adamf\Documents\cosc449-coding\project\office-hours\packages\server\dist\main.js:10926:28)
[server]     at async C:\Users\adamf\Documents\cosc449-coding\project\office-hours\node_modules\@nestjs\core\router\router-execution-context.js:46:28
[server]     at async C:\Users\adamf\Documents\cosc449-coding\project\office-hours\node_modules\@nestjs\core\router\router-proxy.js:9:17 {
[server]   response: {
[server]     statusCode: 400,
[server]     message: 'Invalid login: 535-5.7.8 Username and Password not accepted. For more information, 
go to\n' +
[server]       '535 5.7.8  https://support.google.com/mail/?p=BadCredentials lf11-20020a170902fb4b00b001dccaafe249sm6344747plb.220 - gsmtp',
[server]     error: 'Bad Request'
[server]   },
[server]   status: 400
[server] }
[server]
```

However, despite the error, it seems like it did actually register. Trying to register again told me the email already exists and the account i made appears in the database.

image

**Comment by AdamFipke** [comment]
Also, I can't seem to get the verification emails to send? 

image

It says it was succesfully sent but i'm not getting an email. I set my .env GMAIL_USER and GMAIL_PASSWORD to my second gmail account

**Comment by AdamFipke** [comment]
> Also, I can't seem to get the verification emails to send?

Okay nevermind, I didn't have an app-password set up before. It works now. 

Though still, maybe an error of some kind would be nice saying something like "email service down/not set up"

**Comment by d3li0n** [comment]
> > Also, I can't seem to get the verification emails to send?
> 
> Okay nevermind, I didn't have an app-password set up before. It works now.
> 
> Though still, maybe an error of some kind would be nice saying something like "email service down/not set up"

Intention is whoever deploy production is to have correct cdentials.

**Review by AdamFipke — APPROVED** [review]
Works good. I like it! 

Just left one comment.

Stuff I manually tested:
- creating an account
- creating an account with a valid/invalid student number
- resetting password via forgot password
- login
- verify account

File: packages/app/pages/register.tsx [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+      <Head>
+        <title>Create New Account | HelpMe</title>
+      </Head>
+
+      <Container>
+        <Card className="mx-auto max-w-2xl sm:px-2 md:px-6">
+          <h2 className="my-4 flex items-center text-left">
+            <Tag color="blue" className="uppercase">
+              beta
+            </Tag>
```
[the remark this answers is not in the document]
Comment (d3li0n) [reply packages/app/pages/register.tsx]: Fixed

**Pull request merged**